src/language/completionItemProvider.ts

```diff
@@ -0,0 +1,51 @@
+import {
+  ColumnDefinition,
+  getStandardSQLCompletionProvider,
+  LanguageCompletionProvider,
+  LinkedToken,
+  TableDefinition,
+  TableIdentifier,
+} from '@grafana/experimental';
+import { MACROS } from './macros';
+
+interface CompletionProviderGetterArgs {
+  getDatabases: React.MutableRefObject<() => Promise<TableDefinition[]>>;
+  getTables: React.MutableRefObject<(d?: string) => Promise<TableDefinition[]>>;
+  getColumns: React.MutableRefObject<(database?: string, table?: string) => Promise<ColumnDefinition[]>>;
+}
+
+export const getTimestreamCompletionProvider: (args: CompletionProviderGetterArgs) => LanguageCompletionProvider = ({
+  getDatabases,
+  getTables,
+  getColumns,
+}) => (monaco, language) => {
+  return {
+    // get standard SQL completion provider which will resolve functions and macros
+    ...(language && getStandardSQLCompletionProvider(monaco, language)),
+    triggerCharacters: ['.', ' ', '$', ',', '(', "'"],
+    schemas: {
+      resolve: async () => getDatabases.current(),
+    },
+    tables: {
+      resolve: async (t: TableIdentifier) => {
+        return await getTables.current(t?.schema);
```

Comment:
and I feel like you don't need to await unless you're going to use this promise within this function